### DISCUSSION TITLE
GitHub Runner IP Whitelist?

### DISCUSSION THREAD

**Discussion by wilwade** [ubicloud/ubicloud#1248] [Q&A] [answered]
Are / Is there a way to know the public IP address(es) for the GitHub runners so that the IP can be whitelisted for access / capcha score adjustments / other non-security critical systems?

- I realize I can setup the runner with a VPN as a step if needed (Rather not if it is already a list available)
- I realize that it wouldn't be secure as then others using the same service could also be whitelisted (This about limits not security)

**Comment by fdr** [comment]
> Are / Is there a way to know the public IP address(es) for the GitHub runners so that the IP can be whitelisted for access / capcha score adjustments / other non-security critical systems?
> 
> 
> 
> - I realize I can setup the runner with a VPN as a step if needed (Rather not if it is already a list available)
> 
> - I realize that it wouldn't be secure as then others using the same service could also be whitelisted (This about limits not security)

For the purposes of limits, would it be satisfactory if we relayed to you what the hetzner IP blocks were? We're not guaranteed anything more contiguous. Anything finer grained than that would be very dynamic and increase complication pretty drastically.

**Reply by wilwade** [reply]
That would be good enough for my use cases! Thanks!

**Reply by fdr** [reply, marked as answer]
I think you have two decent options here. The general approach is to obtain a list of IP blocks associated with the Hetzner AS/"autonomous system" BGP advertisements, i.e. how they tell the internet what is to be routed to them and how, if you aren't as familiar with BGP. There are more than few services that kindly offer basic use of these databases for querying free of charge.

The Hetzner AS number is AS24940.

radb.net has a whois server that returns the routes:

```bash
whois  -h whois.radb.net -- '-i origin AS24940' | awk '/^route6?:/ {print $2}'
```

Note you can mess around with that `6?` regular expression if you want no IPv6 prefixes or only IPv6 prefixes: as matches both `route:` (IPv4) and `route6:` fields.

A largely equivalent option is using a HTTPS api from ip.guide that returns JSON:

```bash
curl -sL https://ip.guide/as24940 | jq -r '.routes.v4 | .[]' 
```

This `jq` expression selects and prints only IPv4 routes. It's a bit more involved to get both:

```bash
curl -sL https://ip.guide/as24940  | jq -r '.routes.v4 + .routes.v6 | .[]'
```

As of today, here's the prefixes routed to Hetzner:


```
5.9.0.0/16
5.75.128.0/17
5.161.0.0/17
5.161.0.0/18
5.161.64.0/18
5.161.128.0/18
23.88.0.0/17
37.27.0.0/16
45.15.120.0/22
45.136.70.0/23
45.145.227.0/24
45.148.28.0/22
46.4.0.0/16
49.12.0.0/16
49.13.0.0/16
65.21.0.0/16
65.108.0.0/16
65.109.0.0/16
78.46.0.0/15
78.138.62.0/24
83.219.100.0/22
85.10.192.0/18
88.99.0.0/16
88.198.0.0/16
89.42.83.0/24
89.42.83.0/24
91.107.128.0/17
91.107.128.0/17
91.107.128.0/19
91.190.240.0/21
91.233.8.0/22
94.72.141.0/24
94.130.0.0/16
95.216.0.0/16
95.217.0.0/16
116.202.0.0/16
116.203.0.0/16
128.140.0.0/17
135.181.0.0/16
136.243.0.0/16
138.201.0.0/16
142.132.128.0/17
144.76.0.0/16
148.251.0.0/16
157.90.0.0/16
159.69.0.0/16
162.55.0.0/16
167.233.0.0/16
167.235.0.0/16
168.119.0.0/16
171.25.225.0/24
176.9.0.0/16
178.63.0.0/16
178.212.75.0/24
185.12.65.0/24
185.50.120.0/23
185.107.52.0/22
185.112.180.0/24
185.126.28.0/22
185.136.140.0/23
185.157.83.0/24
185.157.176.0/23
185.157.178.0/23
185.171.224.0/22
185.189.228.0/24
185.189.229.0/24
185.189.230.0/24
185.189.231.0/24
185.209.124.0/22
185.213.45.0/24
185.216.237.0/24
185.226.99.0/24
185.228.8.0/23
185.242.76.0/24
185.253.111.0/24
188.34.128.0/17
188.40.0.0/16
188.245.0.0/16
193.25.170.0/23
193.110.6.0/23
193.163.198.0/24
193.223.77.0/24
194.35.12.0/23
194.42.180.0/22
194.42.184.0/22
194.62.106.0/24
195.60.226.0/24
195.96.156.0/24
195.201.0.0/16
195.248.224.0/24
197.242.84.0/22
201.131.3.0/24
204.29.146.0/24
213.133.96.0/19
213.232.193.0/24
213.239.192.0/18
216.55.108.0/22
217.78.237.0/24
2001:67c:1164::/48
2a01:4f8::/32
2a01:4f9::/32
2a01:4ff:ff01::/48
2a06:1301:4050::/48
2a06:be80::/29
2a0c:de00::/30
2a0e:7700::/32
2a11:e980::/29
2a12:e00::/29
```

This list could probably be made smaller by trying to geolocate the blocks, I was trying to cook something up to do this, but those data sets tend to be somewhat proprietary...or at least, unfortunately, ip.guide lacks some location data that other services have (and perhaps vice versa). Let me know if this simple, yet somewhat large list does the job for you, though.
